### DISCUSSION TITLE
train dreambooth  result

### DISCUSSION THREAD

**Discussion by wen020** [PixArt-alpha/PixArt-alpha#42] [Show and tell]
![0](https://github.com/PixArt-alpha/PixArt-alpha/assets/54690997/ce10adfd-293e-45d2-aa64-cafc9a310d53)

This is the result of training dreambooth. I found that Pixart responds better to prompts, but its consistency with characters is not very good. Has anyone produced better results?

**Comment by apluka34** [comment]
@wen020 I trained with 20 photos, 1600 steps, lr 5e-6, and it produces quite good results. Btw, have you tried to train the text encoder also?
I also want to ask @lawrence-cj for the option to train the text encoder in the script train_dreambooth.py

**Reply by lawrence-cj** [reply]
Actually, I don't think train the text encoder is a good choice. I haven't try this out.

**Reply by apluka34** [reply]
Hmm I think training the text encoder makes it better to adapt a new object (here is sks for example). Or at least add a new token embedding like textual inversion does. What do you think?

**Reply by apluka34** [reply]
I wonder if you have any reasons why training text encoder is not a good choice. Thanks

**Reply by lawrence-cj** [reply]
We have 'sks' in the prompt during training? May I have more explaination about `add a new token embedding like textual inversion does`

**Reply by lawrence-cj** [reply]
> I wonder if you have any reasons why training text encoder is not a good choice. Thanks

Since I haven't seen a way to train text encoders, if any, please show me. Once you train the text encoder, the text features will change, and the previous image-text alignment features will be destroyed. In this case, I don't think training the text encoder is a good choice.

**Reply by apluka34** [reply]
Hmm I quite don't understand. Let me explain it in my words. There is a difference when compared with other diffusion pipelines, that PixArt only use the text encoder as T5-xxl and train the whole transformer with the cross attention. That's why the token embedding of new objects (sks, zwr, ...whatever) seems not important here. 

Do I understand it right?

**Reply by apluka34** [reply]
That's why only training the decoder model (here is transformer) is the best choice?

**Reply by lawrence-cj** [reply]
>  Or at least add a new token embedding like textual inversion does. What do you think?

Any suggestions for choosing a new token?

**Reply by apluka34** [reply]
Hmm I simply think the way textual inversion does is to add the new token embedding. So I don't think it would destroy the image-text alignment. But can you please confirm what I understand is correct or not?

**Reply by apluka34** [reply]
You think that the training of text encoder would destroy the text features. So how dreambooth for example with SD will improve the results instead? Sorry for a lot of questions overthere.

**Reply by lawrence-cj** [reply]
Yeah, I misunderstood. Simply add new token is reasonable. But I don’t really understand how to train text encoder T5 here in dreambooth pipeline.

**Reply by mrT333** [reply]
> [..] Once you train the text encoder, the text features will change, and the previous image-text alignment features will be destroyed. In this case, I don't think training the text encoder is a good choice.

That is what the regularization images in Dreambooth are there for: To avoid catastrophic forgetting / destroying previous image-text alignment. 

Pixart is amazing. I'd love it if it would get a similar community support as Stable diffusion does with thousands of finetunes. I cannot even imagine how the quality of Pixart with its superior training efficiency would look like compared to the SD1.5 community finetunes once people start finetuning it on mass.

But for mass adoption of finetuning (a fully featured, e.g trainable text encoder w reg. images) Dreambooth as a low-barrier option might be crucial as it is way more "idiot proof" than a full finetune which in turn makes it more accessible to the countless hobby users who might not be as proficient in avoiding finetuning mistakes as power users are.

**Reply by lawrence-cj** [reply]
Agree with your suggestion. Is there any repo or paper u think it’s good enough for us to implement the SOTA Dreambooth LoRA.

**Reply by mrT333** [reply]
> Agree with your suggestion. Is there any repo or paper u think it’s good enough for us to implement the SOTA Dreambooth LoRA.

If you are looking for the SOTA Dreambooth LoRA than that is definitely **Hyperdreambooth** ( https://arxiv.org/abs/2307.06949 ) by Google Research. It is an improved version over the original Dreambooth ( https://arxiv.org/abs/2208.12242) in every way, shape and form that was also published by Google Research & Boston University.

HyperDreambooth would have an interesting synergy with PixArt-alpha considering it would have 25x higher training efficiency and 10000x (!) smaller model size (few KB instead of hundeds of MB LORA) than the original Dreambooth that is currently used by everyone in the Stable Diffusion Community.

But the main benefit of HyperDreambooth goes beyond that: Despite its higher training efficiency and tiny size it is just way more capable than the original Dreambooth resulting in achieving better fidelity and retaining more stylistic flexibilty while using only **one** (!) reference image compared to the 10-25 (or in some cases even 40-80 dep. on subject & image quality) reference images necessary for original Dreambooth.

**Reply by mrT333** [reply]
If Hyperdreambooth would be implemented within Pixart it is just a matter of time until it would take over the whole SD community. Let my explain my reasoning:

1) **Shareability** is an important aspect in the frequency on how often LoRAs are used and downloaded, and therefore how high of a popularity they can establish in the community. That is one reason why there are thousands of popular LoRAs compared to only a handful of established full finetunes. Lots of people start to complain about the huge sizes of the SDXL-LoRAs compared to SDv1.5 LoRAs which is i.m.o one of the reason that is hampering SDXLs adoption rate despite its superior quality over base SD1.5. (Although textual inversions are even smaller, they are just not comparable as the fidelity falls too low)

2) The **low number of required reference images** makes Hyperdreambooth more accessible to a huge user base consisting of hobby users boosting usage numbers immensely. One reason why the original Dreambooth is so popular over full finetunes is that it can produce almost the same quality while being less error prone (e.g. preventing catastrophic forgetting through reg. images), needing much less training and a way smaller Dataset. Hyperdreambooth improves over every of those aspects while having higher fidelity in LoRA format than even a full Dreambooth would have (even more so compare to DB LoRA) . 

3) **Higher fidelity and more flexibility** than orig. Dreambooth. LoRAs are often used to introduce a new concept into the model where fidelity is critical, like a TV-show character, a celebrity, anime character etc. (If you don't care about fidelity then it would make more sense to use a textual inversion. And if its not a specific concept but to capture a general style a full finetune would make more sense.) 
If it turns out that characters that are trained through HyperDreambooth LoRAs within Pixart-Alpha have higher resemblance than LoRAs or even full Dreambooths of Stable Diffusion models it is just no competition. People will always prefer the higher fidelity. They don't want to generate images where the character looks "almost as" or "similar to" the intended subject, they want the character to stay true to its identity.

What fascinates me of PixArt-Alpha over SD is on one hand the way higher training efficiency as I expect that to enable the community to create much more powerful finetunes and on the other hand the closer adherence to the prompt as that empowers every single user to create an image closer to their intent. 
I think Hyperdreambooth would supplement both those Pixart-Alpha "superpowers" perfectly with its own higher training efficiency and staying closer to the intent of the artist by retaining higher subject fidelity while still allowing stylistic flexibility.

**Reply by mgyong** [reply]
@quocanh34 Can you share the results of your training dataset and the results of Pixart-dreambooth? so we can compare with what @wen020 did. If you have the code to share that will be great as well. I don't believe @wen020 trained the text encoder during the dreambooth process. Did you @quocanh34 ?

**Comment by lawrence-cj** [comment]
Glad to hear the result. I don't think the consistency issue is caused by PixArt. We also hope for a better solution.

**Comment by attashe** [comment]
Also tried a dreambooth and added a LoRA with PEFT, got some results with default setting, think that need regularization images and more precision prompting during training, also `sks` token hasn't a lot of influence

**Reply by lawrence-cj** [reply]
Nice work. Since we are not experts at dreambooth, can you provide more details about the regularization images and precision prompting? And `sks` hasn't a lot of influence, which means that we don't need to change it to others, no?

**Reply by attashe** [reply]
I meant prior preservation - images for same class generated (or handpicked) by this network without our concept (like some random man when we learning portrait).
There is comparison of results in Using Prior Preservation section https://huggingface.co/blog/dreambooth
For prompting I found this guide: https://rentry.org/59xed3#tags-and-details
Injected lora in following layers: "qkv", "q_linear", "kv_linear", "proj", "t_embedder.mlp.0", "t_embedder.mlp.2", "t_block.1", "fc1", "fc2", "linear". I'm totally not shure that I did right
PS: I'm not expert in dreambooth too, just trained my portrait couple times

**Reply by FurkanGozukara** [reply]
I use real images as regularization images. Prepared 5200 images from unsplash. They are making huge difference. 

man class

![image](https://github.com/PixArt-alpha/PixArt-alpha/assets/19240467/ce492fdb-22d7-4fb7-b7b1-46b62b68ae81)




@attashe are you sure that you have properly trained with accurate captions. I am unsure how captions are used 

it says `data = dict(type='DreamBooth', root='dog6', prompt=['a photo of sks dog'], transform='default_train', load_vae_feat=True)`

so this way we can't provide individual captions for each file?

**Reply by mrT333** [reply]
@attashe since the text encoder is frozen maybe you have more success in avoiding rare tokens like sks or ohwx and rather finetune on a well established token (e.g. a celebrity name that works well prior training and looks similar to your subject if it is a person). 

There are increasing reports that training on well established tokens (e.g. name of a celebrity) can outperform training on rare tokens contrary to the suggestions of the original Dreambooth paper ...as long as the established token is well chosen. 

Rare tokens are chosen to keep concept bleed minimal. But if an established token is well chosen the concept bleed might be beneficial instead of detrimental (e.g. face angles missing in your training dataset might be more accurately supplemented through face angles bleeding in from a similar looking celebrity than drawing from the more general "person" class token that contains all kinds of different face structures which can be very different to your subject). The cost is of course that the well established token is replaced with your subject, which probably should be no problem when you are using a LORA, as you can selectively enable/disable it to get the original or finetuned version of the token.

**Reply by mrT333** [reply]
I also want to confirm the advantages of using of actual real images as class images over AI generated class images as proposed by @FurkanGozukara . 

Just few months ago I came across a paper that did a study on this with the results that on one hand the regularization effect is stronger with real images than it is with AI generated ones, making the final results retain more flexibility over higher training steps and therefore retain higher fidelity until catastrophic forgetting kicks in. On the other hand they also found there are visual artifacts invisible to the human eye that get reinforced every time AI generated images are used for training.  After a couple of "inbreeding" training runs the visual artifacts embedded in the AI generated class images reinforced to a degree that they became visible and distorted the subjects completely. 

I really hope the finetuning community (mostly using stable diffusion currently) would pay more attention to this, as there are so many model merges that on first glance seem to have a high quality but they suffer from catastrophic forgetting making the model lose all its variety and prior knowledge, e.g. making most of the faces look the same.

The reason why that happens is that a lot of finetunes are model merges that contain finetunes trained on AI images that also contain merges of finetunes trained on AI images that also [...] (you get the point) which reduce the regularization effect and in time covertly accumulate visual artifacts and spread those like a covert poison through all model merges until they become visible after a few generations. That would all be avoidable by using real images as class images while further increasing quality & fidelity.

**Reply by attashe** [reply]
@mrT333 will try without rare tokens like 'sks' next weekend, I think you right about this.

**Reply by mgyong** [reply]
@attashe Could you share your results for the Pixart dreambooth example and code? We like to see how it compares with what @wen020 did.

**Reply by mgyong** [reply]
@FurkanGozukara how did create a large regularization dataset for what you did for man but for woman using unsplash? Was it a very tedious process?

**Reply by FurkanGozukara** [reply]
it is beyond tedious. total i spent like 10 days now :D

**Reply by attashe** [reply]
Yes, I trained LoRA with r=8, 3 images, and detailed separated prompts. May be learning rate too high and I need lower and set more steps. In general concept learned, not so much variation, but dataset also very small. 60% generations trash like last image, but others from mediocre to ok.
![photo of sks blue heron on balcony with big teeth and huge beak, high detailed](https://github.com/PixArt-alpha/PixArt-alpha/assets/8243605/7f884233-3957-4612-a9be-54acd2aa185a)
![photo of sks blue heron with big teeth and huge beak, drinking coffee, anime style, high detailed_46547](https://github.com/PixArt-alpha/PixArt-alpha/assets/8243605/f94a8350-2c72-4e1f-9807-934befbb52ac)
![photo of sks blue heron on balcony with big teeth and huge beak, high detailed_189123](https://github.com/PixArt-alpha/PixArt-alpha/assets/8243605/1901298d-5ea6-426d-a0f7-c9ea3b16602e)

**Reply by attashe** [reply]
@mgyong

**Comment by attashe** [comment]
@FurkanGozukara I started rewriting this part for support individual captions. As free alternative used a subset of CelebaHQ as regularization dataset.

**Reply by FurkanGozukara** [reply]
nice

by the way i don't prefer hyper dreambooth. it is lower quality

I have amazing SDXL DreamBooth configuration which allow me to generate superb quality images

I am hoping same can be made here 

![00073-1550383969](https://github.com/PixArt-alpha/PixArt-alpha/assets/19240467/af340de5-9fea-421f-9cff-e1f8a9f66ce2)

![00088-2649357773](https://github.com/PixArt-alpha/PixArt-alpha/assets/19240467/e8c780bd-7527-4700-b679-112e2e2f45e7)

![00119-1713028240](https://github.com/PixArt-alpha/PixArt-alpha/assets/19240467/bbe46579-4050-4a2c-82b7-2e880e32d42f)

**Reply by mrT333** [reply]
how did you come to the conclusion of hyper dreambooth having lower quality than the original dreambooth?

You have to remember the previews in that paper rely on a extremely low reference image size of 1 to showcase the Hyperdreambooths strengths in an extreme scenario. Of course i would never use Hyperdreambooth with just one reference image just the same way I never use the original Dreambooth with just three reference images even though that is how Google advertised it in their original Dreambooth paper.

I saw you used 13 reference images in your Dreambooth training. Even though the number of reference Images is not that high I would still consider your dataset pretty high quality due to the lightning conditions, resolution and most importantly the variety in angle and distance between face and camera. Most hobby users won't use or have access to a dataset of that quality on their subject judging from most LoRAs that I see published in the community.

I would expect that you could reach a similar fidelity to your impressive Dreambooth example above by using a well implemented and well pretrained Hyperdreambooth but using a much lower quality training set than a normal Dreambooth could handle (worse lighting, lower resolution, more gaps in face angles and distances).

**Reply by FurkanGozukara** [reply]
actually my dataset is captured from my phone : poco x3 pro in several minutes :d i didnt even bother to replace clothing 

because i wanted it to be an easy dataset so people can replicate

currently what settings and script you are using to train hyper dreambooth? i can compare with kohya DreamBooth

**Reply by mrT333** [reply]
thats the problem... currently there are only a few poor WIP implementations of Hyperdreambooth available that are not even properly pretrained like the paper suggested.

That is why all we have are the comparisons in the Hyperdreambooth paper which all are done in a extreme scenario of just one reference image and therefor comparing Hyperdreambooth with the original Dreambooth when they both weren't used to their max potential. (Just look at the Dreambooth examples in that paper they are complete garbage as well due to the low reference image count. And your pictures above are proof that the original Dreambooth can do way better than what is shown in that scenario within that paper, same as i would assume the Hyperdreambooth would increase in quality with more reference images)

**Reply by mrT333** [reply]
> actually my dataset is captured from my phone : poco x3 pro in several minutes :d i didnt even bother to replace clothing

Since you are doing a dreambooth of yourself you have control over how the subject is captured and it was quick and easy to change distance and angles. Most people have a limited training data set where they cannot control how the images look like. They have to work with whatever dataset they scrumbled together even if those images always have the same face angle or the same distance. 

To have a good dataset you don't need a professional camera, most important is variety of face angles, distances and high visibility of the face, which I think is exactly what you wanted to demonstrate with your video and succeeded. Good datasets can be made easily & quickly with simple cameras _if you have control over how your subject gets captured_ (which is not always the case).

**Reply by mgyong** [reply]
@FurkanGozukara The example dreambooth images you show here in the thread is done in SDXL right? not using Pixart? Have you been able to get Dreambooth working with Pixart?

**Reply by FurkanGozukara** [reply]
@mgyong they are SDXL . I am yet to train PixArt
